lib/charms/mongodb/v0/config_server_interface.py

```diff
@@ -116,35 +132,74 @@ def __init__(
         )
         # TODO Future PRs handle scale down
 
-    def _on_relation_changed(self, event):
+    def _on_relation_changed(self, event) -> None:
         """Starts/restarts monogs with config server information."""
         relation_data = event.relation.data[event.app]
-        if not relation_data.get(KEYFILE_KEY):
+        if not relation_data.get(KEYFILE_KEY) or not relation_data.get(CONFIG_SERVER_DB_KEY):
             event.defer()
             self.charm.unit.status = WaitingStatus("Waiting for secrets from config-server")
             return
 
-        self.update_keyfile(key_file_contents=relation_data.get(KEYFILE_KEY))
+        updated_keyfile = self.update_keyfile(key_file_contents=relation_data.get(KEYFILE_KEY))
+        updated_config = self.update_config_server_db(
+            config_server_db=relation_data.get(CONFIG_SERVER_DB_KEY)
+        )
+
+        # avoid restarting mongos when possible
+        if not updated_keyfile and not updated_config and self.charm.monogs_initialised:
+            return
+
+        # mongos is not available until it is using new secrets
+        del self.charm.unit_peer_data["mongos_initialised"]
+        logger.info("Restarting mongos with new secrets")
+        self.charm.unit.status = MaintenanceStatus("starting mongos")
+        self.charm.restart_mongos_service()
+
+        # restart on high loaded databases can be very slow (e.g. up to 10-20 minutes).
+        if not self.is_mongos_running():
+            logger.info("mongos has not started, deferring")
+            self.charm.unit.status = WaitingStatus("Waiting for mongos to start")
+            event.defer()
+            return
 
-        # TODO: Follow up PR. Start mongos with the config-server URI
         # TODO: Follow up PR. Add a user for mongos once it has been started
```

Comment:
The user should provided by the Provider. Why do we need user creation on the Requirer side?

Reply:
gotcha